clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2784,16 +2784,80 @@ class SyclOptReportCreator : public SyclKernelFieldHandler {
   }
 };
 
-static CXXMethodDecl *getOperatorParens(const CXXRecordDecl *Rec) {
-  for (auto *MD : Rec->methods()) {
-    if (MD->getOverloadedOperator() == OO_Call)
-      return MD;
+// This function traverses the static call graph from the root of the kernel
+// (e.g. “kernel_parallel_for”) and returns the version of “operator()()” that
+// is called by kernelFunc(). There will only be one call to kernelFunc() in
+// that call graph because the DPC++ headers are structured such that the user’s
+// kernel function is only called once. This ensures that the correct
+// “operator()()” function call is returned, when a named function object used
+// to define a kernel has more than one “operator()()” calls defined in it. For
+// example, in the code below, 'operator()(sycl::id<1> id)' is returned based on
+// the 'parallel_for' invocation which takes a 'sycl::range<1>(16)' argument.
+//   class MyKernel {
+//    public:
+//      void operator()() const {
+//        // code
+//      }
+//
+//      [[intel::reqd_sub_group_size(4)]] void operator()(sycl::id<1> id) const
+//      {
+//        // code
+//      }
+//    };
+//
+//    int main() {
+//
+//    Q.submit([&](sycl::handler& cgh) {
+//      MyKernel kernelFunctorObject;
+//      cgh.parallel_for(sycl::range<1>(16), kernelFunctorObject);
+//    });
+//      return 0;
+//    }
+
+static CXXMethodDecl *
+getCallOperatorInvokedFromKernel(const CXXRecordDecl *KernelFuncObjType,
+                                 FunctionDecl *KernelCallerFunc,
+                                 Sema &SemaRef) {
+
+  CallGraph SYCLCG;
+  SYCLCG.addToCallGraph(SemaRef.getASTContext().getTranslationUnitDecl());
+
+  // This code returns the 'lambda' call operator.
+  if (KernelFuncObjType->isLambda()) {
+    for (auto *MD : KernelFuncObjType->methods()) {
+      if (MD->getOverloadedOperator() == OO_Call)
+        return MD;
+    }
   }
+
+  // This code returns the functor's call operator.
+  if (KernelCallerFunc && KernelCallerFunc->hasBody() &&
+      KernelCallerFunc->hasAttr<SYCLKernelAttr>()) {
+
+    CallGraphNode *KernelCallerFuncNode = SYCLCG.getNode(KernelCallerFunc);
+    CXXMethodDecl *OperatorCall = nullptr;
+
+    // Iterate through each funtion invoked from the kernel root, find the
+    // function call operator and make sure it is a member of the kernel fuctor.
+    for (const CallGraphNode *ChildNode : *KernelCallerFuncNode) {
+      if (auto *Callee = dyn_cast<CXXMethodDecl>(ChildNode->getDecl())) {
+        Callee = Callee->getMostRecentDecl();
+        if (Callee->getParent() == KernelFuncObjType &&
+            Callee->isCXXClassMember() &&
```

Comment:
Is check `Callee->isCXXClassMember()` necessary?

Reply:
I added this as an additional check to ensure that the `operator()()` function is a member of the functor class. I can remove it since `getParent()` returns the parent class in which `operator()()` method is defined.